### ISSUE TITLE
8.2 height calculation error

### ISSUE THREAD

**Issue by pwilkin** [vaadin/framework#10185]
HorizontalLayouts which have no components in them have non-zero height, but apparently aren't included in height calculation, which causes issues with calculating the height of expanding components. 

Sample: 

```
public class LayoutCalculation extends VerticalLayout {

    public LayoutCalculation() {
        setSizeFull();
        addComponent(new HorizontalLayout());
        Panel p = new Panel();
        p.setSizeFull();
        addComponent(p);
        setExpandRatio(p, 1.0f);
        addComponent(new HorizontalLayout());
        addComponent(new Button("TEST"));
    }

}
```

The button will be rendered partially outside the screen.

**Comment by Legioth** [comment]
This happens because `VerticalLayout` looks at the height of the child component instead of the height of the slot. The height of the HorizontalLayout itself is zero, but the height of the wrapping slot is non-zero because its child has `display: inline-block` and therefore `line-height` is taken into account. There would be a similar problem also in cases when the HorizonalLayout would be non-empty but still have a height that is less than the `lineHeight`.

This means that there are actually two problems: 
1. A child with a low height still occupies space. This might be fixed by globally by setting `lineHeight: 0` for `v-slot`, but that might cause issues related to e.g. https://github.com/vaadin/framework/issues/3364. A less risky approach might be that HorizontalLayout would give itself e.g. `display: none` when it has no children, although that might also have some unintended side effects (e.g. disappearing spacing).
2. The height of the slot is not taken into account. It's conceivable that the slot might have height other than the child component also for other reasons than the current `lineHeight` issue. This might be fixed by changing how resize listeners are attached by `AbstractOrderedLayoutConnector`, but this approach also has some risks of unintended side effects.

**Comment by Legioth** [comment]
And by the way, starting from Vaadin Framework 8.0, you can use the `addComponentsAndExpand` method instead of `addComponent`+ `setSizeFull`+ `setExpandRatio`, thus significantly simplifying the example code here.

**Comment by pwilkin** [comment]
At this point, I use the workaround in point 1 (server-side, I just do `setVisible(getComponentCount() > 0)`). I'd say this might be a viable solution (isn't spacing realized by a separate div anyways?)

**Comment by Legioth** [comment]
There won't be any difference in most cases, but there's still a semantic difference between an empty layout and a hidden layout. One such case is if the layout has a caption: it should be shown even if the layout is empty, but not if it's hidden.

**Comment by pwilkin** [comment]
Hmm, how about
```
.v-slot:empty {
 line-height: 0 !important;
}
```

?

**Comment by Legioth** [comment]
I don't think that would work either since the slot isn't empty, it contains a `HorizontalLayout`. What might work is something like this:
```
.v-horizontallayout:empty {
  display:none;
}
```

This approach isn't perfect either since a HorizontalLayout with spacing enabled should still occupy the size of the spacing.

**Comment by mvysny** [comment]
I don't mind an empty visible `HorizontalLayout` having non-zero height (exactly as it is rendered now). The logic would be that even though it's empty, it's still a component occupying a slot (even though it has zero children and hence draws nothing). This logic follows the same pattern for an empty `Label` (shows nothing but still occupies space).

If we accept this logic, then the fix would IMHO simply be to fix the layouting engine to correctly consider  empty `HorizontalLayout` as having a non-zero height.

**Comment by mvysny** [comment]
The bug is still present in Vaadin 8.5.2. Attaching a sample project which reproduces the issue. Just run `./gradlew appRun`.
[bug10185.zip](https://github.com/vaadin/framework/files/2446328/bug10185.zip)
